services/supervisor/src/migration_state.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright 2020 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use exonum::{
+    crypto::Hash,
+    runtime::{versioning::Version, ExecutionError},
+};
+use exonum_derive::*;
+use exonum_proto::ProtobufConvert;
+use serde_derive::{Deserialize, Serialize};
+
+use super::{proto, AsyncEventState, MigrationError};
+
+/// State of a migration.
+#[derive(Debug, Clone)]
+#[derive(ProtobufConvert, BinaryValue)]
+#[derive(Serialize, Deserialize)]
+#[protobuf_convert(source = "proto::MigrationState")]
+pub struct MigrationState {
+    /// Migration process state.
+    #[serde(rename = "state")]
+    pub inner: AsyncEventState,
+
+    /// Current artifact data version.
+    #[protobuf_convert(with = "exonum::helpers::pb_version")]
+    pub version: Version,
+
+    /// Expected state hash. Equals to the first obtained local migration state hash.
+    /// For a good scenario, all the hashes should be equal between each other.
+    /// For the bad scenario, at least one node obtains the different hash and that's enough
+    /// to consider migration failed.
+    #[protobuf_convert(with = "exonum::helpers::pb_optional_hash")]
+    #[serde(skip)]
+    pub(crate) expected_state_hash: Option<Hash>,
```

Comment:
Nit: maybe, call this field `reference_state_hash`? IMO, `expected` implies that this hash is *always* correct, which isn't the case.